### DISCUSSION TITLE
An example of creating a canvas overlay without a clipPath and hiding an object or part of an object outside the specified area. Example of aligning an object to the canvas

### DISCUSSION THREAD

**Discussion by i-prikot** [fabricjs/fabric.js#7166] [Show and tell]
Fully opaque or translucent overlay, which is applied over the canvas at the end of rendering https://jsfiddle.net/i_prikot/u582zkej/

**Comment by i-prikot** [comment]
https://user-images.githubusercontent.com/21138391/123809603-f1dbf280-d91b-11eb-9522-75edab7dd721.mp4

**Reply by melchiar** [reply]
This is great!

**Reply by i-prikot** [reply]
Thanks! )

**Comment by asturur** [comment]
this is good, is what a lot people ask for. We can make it a demo for sure.

**Comment by mauricekindermann** [comment]
For anyone stumbling on this in the future - the after:render approach is brilliant but caused flickering for us with semi-transparent overlays (the overlay would visually double-up on rapid re-renders like clicks/selections). Our solution was a separate static HTML canvas element positioned over the FabricJS canvas (pointer-events: none) that only redraws on zoom/pan changes - same even-odd path winding technique, just on its own canvas layer that doesn't re-render on every fabric interaction.

<img width="568" height="354" alt="image" src="https://github.com/user-attachments/assets/8493e0ce-c87e-449f-bb59-4ce6b043f49f" />

**Reply by asturur** [reply]
There shouldn't be any flickering, the after:render is syncronous, that is very strange.

**Reply by mauricekindermann** [reply]
The flicker was likely my own event handling - I was trying to adjust opacity on selection/cleared.

I ended up re-writing it in the end with a solid color using a single canvas, and instead override _renderOverlay to inject the overlay into Fabric's existing render cycle. Works well - the overlay draws after Fabric's native overlays but before selection controls, so handles/borders/selection show up correctly on top:

```javascript
const originalRenderOverlay = canvas.instance._renderOverlay.bind(canvas.instance);

canvas.instance._renderOverlay = function(ctx) {
    originalRenderOverlay(ctx);

    ctx.save();
    ctx.beginPath();
    ctx.rect(0, 0, cw, ch);
    ctx.transform(...vpt);
    ctx.rect(paperX, paperY, paperWidth, paperHeight);
    ctx.fillStyle = overlayColor;
    ctx.fill('evenodd');
    ctx.restore();
};
```

<img width="500" height="354" alt="image" src="https://github.com/user-attachments/assets/26514785-9819-4413-a974-713ea16c8d44" />
